File: lib/risk_engine.dart
import 'package:cloud_firestore/cloud_firestore.dart';


class RiskResult {
  final String preeklampsi;
  final String diyabet;
  final String preterm;

  RiskResult({
    required this.preeklampsi,
    required this.diyabet,
    required this.preterm,
  });
}

class RiskEngine {
  static Future<String> calculatePreeklampsi({
    required String uid,
    required int sistolik,
    required int diastolik,
    required bool gormeBozuklugu,
    required bool basAgrisi,
    required bool sislik,
    required bool chronicHypertension,
  }) async {
    if (sistolik >=160 || diastolik >= 110 ){
      return "HIGH";
    }
    int score = 0;
    if (sistolik >=140) score += 2;
    if (diastolik >= 90) score += 2;
    if (basAgrisi) score += 1;
    if (gormeBozuklugu) score += 1;
    if (sislik) score += 1;
    if (chronicHypertension) score += 2;

    String risk;

    if (score <=2){
      risk = "LOW";
    }else if (score <= 5){
      risk = "MEDIUM";
    }else {
      risk = "HIGH";
    }

    final query = await FirebaseFirestore.instance
      .collection("risk_olcumleri")
      .where("uid", isEqualTo: uid)
      .orderBy("tarih", descending: true)
      .limit(3)
      .get();
    if (query.docs.length == 3){
      int abnormalCount = 0;

      for (var doc in query.docs){
        final data = doc.data() as Map<String, dynamic>;
        final s = data["sistolik"] ?? 0;
        final d = data["diastolik"] ?? 0;

        if (s >= 140 || d>= 90){
          abnormalCount++;
        }
      }
      if (abnormalCount == 3){
        if (risk == "LOW") risk = "MEDIUM";
        if(risk == "MEDIUM") risk = "HIGH";
      }
    }
    return risk;
  }
  static String calculateDiyabet({
    required double? aclik,
    required double? tokluk,
    required bool asiriSusama,
    required bool sikIdrar,
    required bool diabetes,
  }) {
    if ((aclik ?? 0) >= 126 || (tokluk ?? 0) >= 200){
      return "HIGH";
    }
    int score = 0;
    if ((aclik ?? 0) >= 100) score +=2;
    if ((tokluk ?? 0) >= 140) score +=2;
    if (asiriSusama) score +=1;
    if (sikIdrar) score +=1;
    if (diabetes) score +=2;

    if (score <=2) return "LOW";
    if (score >2 && score <=5) return "MEDIUM";
    return "HIGH";
  }
  static String calculatePreterm({
    required bool karinKasilma,
    required bool akinti,
    required bool belAgrisi,
    required double stresSeviyesi,
    required bool previousPreterm,
    required bool multiplePregnancy,
  }){
    int score = 0;

    if (karinKasilma) score +=2;
    if (akinti) score +=1;
    if (belAgrisi) score +=1;
    if (stresSeviyesi >= 4) score +=1;
    if (previousPreterm) score +=2;
    if (multiplePregnancy) score +=2;

    if (score <= 2) return "LOW";
    if(score >2 && score<=5) return "MEDIUM";
    return "HIGH";
  }
  static Future<void> sendRiskNotification({
    required String uid,
    required String riskType,
    required String riskLevel,
  }) async {

    if (riskLevel != "HIGH") return;

    final userDoc = await FirebaseFirestore.instance
        .collection("users")
        .doc(uid)
        .get();

    final userData = userDoc.data();
    final doctorId = userData?["assignedDoctor"];
    final patientName = userData?["name"] ?? "Bir hasta";

    // Aynı risk için daha önce bildirim var mı kontrol et
    final existing = await FirebaseFirestore.instance
        .collection("notification")
        .where("uid", isEqualTo: uid)
        .where("type", isEqualTo: "risk_alert")
        .where("riskType", isEqualTo: riskType)
        .limit(1)
        .get();

    if (existing.docs.isNotEmpty) {
      return;
    }

    // Hamile kullanıcıya bildirim
    await FirebaseFirestore.instance.collection("notification").add({
      "uid": uid,
      "type": "risk_alert",
      "riskType": riskType,
      "title": "Risk Uyarısı",
      "message":
      "$riskType riski yüksek tespit edildi. Lütfen doktorunuzla iletişime geçiniz.",
      "isRead": false,
      "createdAt": FieldValue.serverTimestamp(),
    });

    // Doktora bildirim
    if (doctorId != null) {
      await FirebaseFirestore.instance.collection("notification").add({
        "uid": doctorId,
        "type": "risk_alert",
        "riskType": riskType,
        "title": "Riskli Hasta",
        "message": "$patientName adlı hastada $riskType riski HIGH çıktı.",
        "isRead": false,
        "createdAt": FieldValue.serverTimestamp(),
      });
    }
  }
}

File: web_dashboard/riskEngine.js
import {
  collection,
  query,
  where,
  orderBy,
  limit,
  getDocs,
  doc,
  getDoc,
  setDoc,
  addDoc
} from "https://www.gstatic.com/firebasejs/12.12.0/firebase-firestore.js";

function getDB() {
  return window.db;
}
export const RiskLevel = {
  LOW: "LOW",
  MEDIUM: "MEDIUM",
  HIGH: "HIGH"
};

export async function calculatePreeklampsi({
  uid,
  sistolik,
  diastolik,
  basAgrisi,
  gorme,
  sislik,
  chronic = false
}) {

  if (sistolik >= 160 || diastolik >= 110) {
    return RiskLevel.HIGH;
  }

  let score = 0;

  if (sistolik >= 140) score += 2;
  if (diastolik >= 90) score += 2;
  if (basAgrisi) score += 1;
  if (gorme) score += 1;
  if (sislik) score += 1;
  if (chronic) score += 2;

  let risk = RiskLevel.LOW;

  if (score <= 2) risk = RiskLevel.LOW;
  else if (score <= 5) risk = RiskLevel.MEDIUM;
  else risk = RiskLevel.HIGH;

  const q = query(
    collection(getDB(), "risk_olcumleri"),
    where("uid", "==", uid),
    orderBy("createdAt", "desc"),
    limit(3)
  );

  const snap = await getDocs(q);

  if (snap.docs.length === 3) {
    let abnormal = 0;

    snap.forEach(doc => {
      const d = doc.data();
      if (d.sistolik >= 140 || d.diastolik >= 90) {
        abnormal++;
      }
    });

    if (abnormal === 3) {
      if (risk === "LOW") risk = "MEDIUM";
      else if (risk === "MEDIUM") risk = "HIGH";
    }
  }

  await setDoc(doc(db, "users", uid), {
    riskLevel: risk.toLowerCase()
  }, { merge: true });

  return risk;
}

export function calculateDiyabet({
  aclik,
  tokluk,
  susama,
  idrar,
  diabetes = false
}) {

  if ((aclik >= 126) || (tokluk >= 200)) {
    return RiskLevel.HIGH;
  }

  let score = 0;

  if (aclik >= 100) score += 2;
  if (tokluk >= 140) score += 2;
  if (susama) score += 1;
  if (idrar) score += 1;
  if (diabetes) score += 2;

  if (score <= 2) return RiskLevel.LOW;
  if (score <= 5) return RiskLevel.MEDIUM;
  return RiskLevel.HIGH;
}

export function calculatePreterm({
  kasilma,
  akinti,
  bel,
  stres,
  prev = false,
  multi = false
}) {

  let score = 0;

  if (kasilma) score += 2;
  if (akinti) score += 1;
  if (bel) score += 1;

  if (stres === 5) score += 3;
  else if (stres >= 4) score += 2;

  if (prev) score += 2;
  if (multi) score += 2;

  if (score <= 2) return RiskLevel.LOW;
  if (score <= 5) return RiskLevel.MEDIUM;
  return RiskLevel.HIGH;
}